Add texture coordinate type to RenderStates

cschreib opened the pull request:
* [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before? (see https://github.com/SFML/SFML/issues/1773)
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [x] Have you provided some example/test code for your changes?
* [x] If you have additional steps which need to be performed list them as tasks!

----

## Description

This pull request adds support for changing the texture coordinate type in `RenderStates`, so that `VertexArray` and `VertexBuffer` can be used with normalized texture coordinates. Other drawables still enforce texture coordinates in pixels. The rationale for this modification is provided in #1773.

This change is fully backward compatible. However, to keep the API clean and reduce header inclusion dependency, I decided to move the `CoordinateType` enum out of `sf::Texture` and into its own file. In the process, I deprecated using the `sf::Texture::CoordinateType` in `sf::Texture::bind()` and added an overload that uses `sf::CoordinateType`. I thought this would be neater in the long run, but if you think it's fine to just use `sf::Texture::CoordinateType` in `RenderStates` (which forces including the `Texture.hpp` header), and avoid the deprecation, I can revert that.

This PR is related to the issue #1773.

## Tasks

* [x] Tested on Linux
* [ ] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

```cpp
#include <SFML/Graphics.hpp>

int main()
{
    sf::RenderWindow window(sf::VideoMode(sf::Vector2u{1280, 720}), "Minimal, complete and verifiable example");
    window.setFramerateLimit(60);

    sf::Texture texture;
    bool        success = texture.loadFromFile("../examples/assets/logo.png");

    // 'v1/s1' uses pixel coordinates
    sf::VertexArray v1(sf::PrimitiveType::Triangles, 6);
    v1[0].position = {0, 0};
    v1[1].position = {256, 0};
    v1[2].position = {256, 256};
    v1[3].position = {256, 256};
    v1[4].position = {0, 256};
    v1[5].position = {0, 0};

    v1[0].texCoords = {0, 0};
    v1[1].texCoords = {512, 0};
    v1[2].texCoords = {512, 512};
    v1[3].texCoords = {512, 512};
    v1[4].texCoords = {0, 512};
    v1[5].texCoords = {0, 0};

    sf::RenderStates s1;
    s1.texture        = &texture;
    s1.coordinateType = sf::CoordinateType::Pixels; // default

    // 'v2/s2' uses normalised coordinates
    sf::VertexArray v2(sf::PrimitiveType::Triangles, 6);
    v2[0].position = {256, 0};
    v2[1].position = {512, 0};
    v2[2].position = {512, 256};
    v2[3].position = {512, 256};
    v2[4].position = {256, 256};
    v2[5].position = {256, 0};

    v2[0].texCoords = {0, 0};
    v2[1].texCoords = {1, 0};
    v2[2].texCoords = {1, 1};
    v2[3].texCoords = {1, 1};
    v2[4].texCoords = {0, 1};
    v2[5].texCoords = {0, 0};

    sf::RenderStates s2;
    s2.texture        = &texture;
    s2.coordinateType = sf::CoordinateType::Normalized;

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        window.clear();

        window.draw(v1, s1);
        window.draw(v2, s2);

        window.display();
    }
}
```

eXpl0it3r commented:
Thanks a lot for this addition! 🙂 

We've already seen a lot of points from @binary1248 in the linked issue, but I'd like to hear some more thoughts from others. If you can bear with us a bit longer, we hopefully get this reviewed soon-ish. 😉

Bromeon commented on include/SFML/Graphics/RenderStates.hpp:
```diff
@@ -76,6 +77,14 @@ class SFML_GRAPHICS_API RenderStates
+    /// \param theCoordinateType Texture coordinate type to use
+    ///
+    ////////////////////////////////////////////////////////////
+    RenderStates(CoordinateType theCoordinateType);
```
Is this constructor actually useful or is it there just to be consistent with the other ones?

cschreib replied:
I only added it for consistency; I doubt it would be of much use on its own without providing also a texture. Would you rather remove it?

Bromeon replied:
I tend to have the mindset "if it's not useful, don't add it" as we can always add things later, but not remove them. How do others see this?

Bromeon commented on include/SFML/Graphics/CoordinateType.hpp:
```diff
@@ -0,0 +1,46 @@
+/// \see sf::Texture::bind
+///
+////////////////////////////////////////////////////////////
+enum CoordinateType
+{
+    Normalized, //!< Texture coordinates in range [0 .. 1]
+    Pixels      //!< Texture coordinates in range [0 .. size]
+};
```
This spills the constants to the `sf` namespace, i.e. `sf::Normalized` and `sf::Pixels`. Is that what we want (like with [`PrimitiveType`](https://www.sfml-dev.org/documentation/2.5.1/group__graphics.php#ga5ee56ac1339984909610713096283b1b))?

cschreib replied:
The right thing would be to use C++11 `enum class`, but I think that's not an option for SFML. To prevent pollution, we could do something like
```c++
namespace Coordinate
{
    enum Type
    {
        Normalized,
        Pixels
    };
}
```

That will still allow implicit conversions, which `enum class` wouldn't allow, but it's less of an issue. I have no objection going either way.

cschreib replied:
Or, with more boilerplate ([kind of like this](https://musingstudio.com/2014/05/20/how-to-emulate-c11s-enum-class-in-c03/)):
```c++
class CoordinateType
{
public:

    static const CoordinateType Normalized;
    static const CoordinateType Pixels;

    CoordinateType() {}
    explicit CoordinateType(int theValue) : value(theValue) {}
    operator int() const { return value; }

private:
 
    int value;
};
```

But this may start to become a bit obscure.

Bromeon replied:
I wouldn't overdo it, both the simple `enum` and the `Coordinate` namespace are good for me. Since this will likely become a C++11 `enum class` in SFML 3, we can also keep things as is for now. I was more asking to make sure everyone is OK with the symbols being in `sf` namespace.

Bromeon commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -52,6 +53,9 @@ class SFML_GRAPHICS_API Texture : GlResource
     ////////////////////////////////////////////////////////////
     /// \brief Types of texture coordinates that can be used for rendering
     ///
+    /// This enum is present for backward compatibility only. Prefer using
+    /// sf::CoordinateType instead.
+    ///
```
What about changing this `enum` to a `typedef` referring to `sf::CoordinateType`?
Then we don't have to deprecate `bind()`, just this typedef (not sure if possible with `SFML_DEPRECATED`).

cschreib replied:
That is what I did at first, and discovered that indeed the deprecation propagates to the typedef too :(

Bromeon replied:
I would still do it, we can also just mention "deprecated" in the docs.
There was no standard deprecation mechanism in C++ until recently, so we have to live with the options available...

cschreib replied:
I've re-targeted this PR against `master` now, so I just removed the old function since we don't need to preserve backward compatibility in 3.0.

Bromeon commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -569,7 +573,16 @@ class SFML_GRAPHICS_API Texture : GlResource
     ///
     ////////////////////////////////////////////////////////////
-    static void bind(const Texture* texture, CoordinateType coordinateType = Normalized);
+    static void bind(const Texture* texture, sf::CoordinateType coordinateType = sf::CoordinateType::Normalized);
```
`sf::CoordinateType::Normalized` is C++11, we should use `sf::Normalized`.

cschreib replied:
My C++03 is so rusty! Fixed.

Bromeon replied:
🦀

Bromeon commented on src/SFML/Graphics/RenderStates.cpp:
```diff
@@ -41,61 +41,91 @@ const RenderStates RenderStates::Default(BlendMode(
-shader   (NULL)
+blendMode     (theBlendMode),
+transform     (),
+coordinateType(CoordinateType::Pixels),
```
C++11, see above

cschreib replied:
My C++03 is so rusty! Fixed.

Bromeon commented on src/SFML/Graphics/Texture.cpp:
```diff
@@ -742,7 +742,7 @@ void Texture::bind(const Texture* texture, CoordinateType coordinateType)
 
         // Check if we need to define a special texture matrix
-        if ((coordinateType == Pixels) || texture->m_pixelsFlipped)
+        if ((coordinateType == sf::Pixels) || texture->m_pixelsFlipped)
```
In case we go for the typedef, this change could be reverted.

cschreib replied:
Done

Bromeon reviewed (commented):
Thanks for the PR!
Added a few comments and questions 🙂

cschreib commented:
Thank you for the comments @Bromeon, I have replied to all of them.

cschreib commented:
The last round of comments from @Bromeon was asking for opinion from other maintainers, but there has been no reply since. Should we still wait for these comments, or should I just go ahead an implement the changes suggested by @Bromeon?

cschreib commented:
I have re-targeted this PR against master, hence:
 - Using C++17 to improve the code.
 - Remove functions without care for breaking backward compatibility.

This addresses all the remaining comments from @Bromeon.

Bambo-Borris commented on include/SFML/Graphics/CoordinateType.hpp:
```diff
@@ -0,0 +1,46 @@
+//
+////////////////////////////////////////////////////////////
+
+#ifndef SFML_COORDINATETYPE_HPP
```
Pretty sure SFML master now uses `#pragma once` instead of the C era header guards, worth changing to be consistent with the rest of master.

cschreib replied:
Indeed, done. I also noticed the copyright year was out of date (this PR is old!), so I fixed that too.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

cschreib commented:
If there are no further comments on this PR and the CI passes, are there any remaining blockers for it be merged? I have been maintaining this branch for close to two years now :)

eXpl0it3r commented on src/SFML/Graphics/Shape.cpp:
```diff
@@ -192,6 +192,7 @@ void Shape::draw(RenderTarget& target, const RenderStates& states) const
     RenderStates statesCopy(states);
 
     statesCopy.transform *= getTransform();
+    statesCopy.coordinateType = CoordinateType::Pixels;
```
Could the state not be overwritten? when calling draw?

cschreib replied:
Hmm there are two ways I can understand your question, but in both cases the answer is no :)
 - If you mean "the state could be overwritten in the `target.draw()` call", then no because `RenderTarget::draw()` takes the state as a const reference, and does not modify it.
 - If you mean "we should not overwrite the state in Shape::draw()", then I'd say we have to, because the interface of `Shape` imposes integer texture coordinates (see `Shape::setTextureRect()`, `Shape::updateTexCoords()`). So the class would break if a user asked for different coordinate type. We could check and assert instead of silently overwrite though, but I'm not sure it's best.

eXpl0it3r replied:
What happens if I call `window.draw(shape, sf::CoordinateType::Normalized)`?

cschreib replied:
As currently implemented, this is equivalent to `window.draw(shape)`. It calls `Shape::draw()` which overrides the coordinate type to `Pixels`.

It's probably an error to do this, because `Shape` in its current form cannot work with normalized coordinates. There's nothing fundamental about it, we could update all SFML classes to also support normalized coordinates. I left this out of this PR because there wasn't demand for it (in the issue https://github.com/SFML/SFML/issues/1773 it was made clear that pixel coordinates must remain the default), and also to keep the PR focused on the core capability.

eXpl0it3r replied:
Hmm, I guess that makes sense. I wonder though whether we should have an assert if someone does choose to pass normalized as type 🤔

cschreib replied:
Thinking about this some more; I'm not sure the assert or `sf::err` is necessary, because there is already a precedent for silently overwriting the render state without warning. Indeed this is already done today for the texture:
https://github.com/SFML/SFML/blob/1e9161b57e475764e91e1efd9bde80102f16618c/src/SFML/Graphics/Shape.cpp#L197
https://github.com/SFML/SFML/blob/1e9161b57e475764e91e1efd9bde80102f16618c/src/SFML/Graphics/Shape.cpp#L203
https://github.com/SFML/SFML/blob/1e9161b57e475764e91e1efd9bde80102f16618c/src/SFML/Graphics/Sprite.cpp#L138
https://github.com/SFML/SFML/blob/1e9161b57e475764e91e1efd9bde80102f16618c/src/SFML/Graphics/Text.cpp#L374

If someone were to call `window.draw(sprite, someOtherTexture)`, the call would ignore `someOtherTexture` and just use the texture set up inside `sprite`, without warning. So it seems there is already an implicit contract that render states may be overwritten inside `draw()`, if it is necessary to display the `Drawable` as intended.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

eXpl0it3r reviewed (approved):
Sorry for all the delays! 😔

Looks good to me!

ChrisThrasher commented on include/SFML/Graphics/CoordinateType.hpp:
```diff
@@ -0,0 +1,43 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2023 Laurent Gomila (laurent@sfml-dev.org)
```
```suggestion
// Copyright (C) 2007-2024 Laurent Gomila (laurent@sfml-dev.org)
```

eXpl0it3r commented:
Thank you for creating this PR and keep it alive over time! 🙂
